sage.py

```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    services = [alert[3] for window in alert_seq_epi for alert in window]
+    unique_signatures = list(set([alert[4] for window in alert_seq_epi for alert in window]))
+    events = [len(window) for window in alert_seq_epi]
+    alert_volume = round(sum(events) / float(len(events)), 1)
+
+    # Make exact start/end times based on alert timestamps
+    timestamps = [alert[2] for window in alert_seq_epi for alert in window]
+    first_ts, last_ts = min(timestamps), max(timestamps)
+
+    # Make the start/end times the actual elapsed times
+    start_time = (first_ts - start_times[tid]).total_seconds()
+    end_time = (last_ts - start_times[tid]).total_seconds()
+    period = end_time - start_time
+
+    episode = (start_time, end_time, mcat, len(events), alert_volume, period, services, unique_signatures, (first_ts, last_ts))
+    return episode
+
+
+def _legend_without_duplicate_labels(ax, fontsize=10, loc='upper right'):
+    handles, labels = ax.get_legend_handles_labels()
+    unique = [(h, l) for i, (h, l) in enumerate(zip(handles, labels)) if l not in labels[:i]]
+    unique = sorted(unique, key=lambda x: x[1])
+    ax.legend(*zip(*unique), loc=loc, fontsize=fontsize)
+
+
+def _plot_alert_volume_per_episode(tid, attacker_victim, host_episodes, mcats):
+    plt.figure(figsize=(10, 10))
+    ax = plt.gca()
+    plt.title('Micro attack episodes | Team: ' + str(tid) + ' | Host: ' + '->'.join(attacker_victim))
+    plt.xlabel('Time Window (sec)')
+    plt.ylabel('Micro attack stages')
+    # NOTE: Line thicknesses are on per-host basis
+    tmax = max([epi[4] for epi in host_episodes])
+    tmin = min([epi[4] for epi in host_episodes])
+    for idx, ep in enumerate(host_episodes):
+        xax = list(np.arange(ep[0], ep[1] + 1))
+        yax = [mcats.index(ep[2])] * len(xax)
+        thickness = ep[4]
+        lsize = ((thickness - tmin) / (tmax - tmin)) * (5 - 0.5) + 0.5 if (tmax - tmin) != 0.0 else 0.5
+        # lsize = np.log(thickness) + 1 TODO: Either take log or normalize between [0.5 5]
+        msize = (lsize * 2) + 1
+        ax.plot(xax, yax, color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], linewidth=lsize)
+        ax.plot(ep[0], mcats.index(ep[2]), color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], marker='.', linewidth=0,
+                markersize=msize, label=micro2macro[micro[ep[2]]])
+        ax.plot(ep[1], mcats.index(ep[2]), color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], marker='.', linewidth=0,
+                markersize=msize)
+        plt.yticks(range(len(mcats)), [x.split('.')[1] for x in micro.values()], rotation=0)
+    _legend_without_duplicate_labels(ax)
+    plt.grid(True, alpha=0.4)
+
+    # plt.tight_layout()
+    # plt.savefig('Pres-Micro-attack-episodes-Team'+str(tid) +'-Connection'+ attacker[0]+'--'+attacker[1]+'.png')
+    plt.show()
+
+
+# Step 2: Create alert sequence and get episodes
+def aggregate_into_episodes(_team_alerts, step=150):
```

Comment:
PEP was often complaining about local variables/parameters shadowing the global ones (which I pretty much agree with), so I have prefixed these local variables/parameters with an underscore